### PR TITLE
IN-928 Restructure the config module to use a 'Config' class

### PR DESCRIPTION
**Pull request by jonavellecuerdo**
### What does this PR do?

Refactor config module to use a `Config` class to improve code readability and flow.
The CLI `main()` method was also updated to exit cleanly in the event of failed
connection tests.

### Helpful background context

* **Updating the CLI `main()` method**
   * During testing, it was discovered that the `Traceback` was printing twice in the CloudWatch logs.
   * In order to address this issue, the following updates were made: 
      * Wrap each connection test in a `try-except` block;
      * Moving the wrapped code into a function in `carbon.app.run_all_connection_tests()` to keep `cli.py` 
   organized.
      * I debugged a couple of errors with @ghukill and he suggested that "logging.exception [may] not [be] needed for an understood exception like `oracledb.exceptions.DatabaseError: ORA-12170: TNS:Connect timeout occurred`" and posted that `logging.error` is sufficient for such cases (this requires ignoring Ruff rule `TRY400`).
   * I originally tried to store this method in `carbon.app.helpers`, but it wasn't possible due to a circular import.

* **Added a `.dockerignore` file.**
   * Copied the `.dockerignore` file from the Python CLI template repo!

### How can a reviewer manually see the effects of these changes?

Review the following CloudWatch logs in `Stage-Workloads`.

1. [Log](https://us-east-1.console.aws.amazon.com/cloudwatch/home?region=us-east-1#logsV2:log-groups/log-group/carbon-ecs-stage/log-events/carbon$252Fcarbon-ecs-stage$252F89cd857cf60b451e8374aa10fb7b0ff1) showing clean exit from `carbon.app.run_all_connection_tests` and simplified error log for failed Data Warehouse connection test.
   image

2. [Log](https://us-east-1.console.aws.amazon.com/cloudwatch/home?region=us-east-1#logsV2:log-groups/log-group/carbon-ecs-stage/log-events/carbon$252Fcarbon-ecs-stage$252Facc3ed6069de42c9ab8c0aa0c1c3b94c) showing clean exit from `carbon.app.run_all_connection_tests` and simplified error log for failed Elements FTP connection test.
   image

3. [Log](https://us-east-1.console.aws.amazon.com/cloudwatch/home?region=us-east-1#logsV2:log-groups/log-group/carbon-ecs-stage/log-events/carbon$252Fcarbon-ecs-stage$252F9d13654eb90e409b80063548605d3a14) showing clean exit from `carbon.app.run_all_connection_tests` where all connection tests were successful.
   image

4. [Log](https://us-east-1.console.aws.amazon.com/cloudwatch/home?region=us-east-1#logsV2:log-groups/log-group/carbon-ecs-stage/log-events/carbon$252Fcarbon-ecs-stage$252F125632ba37244379a7b33e2ecbe8cce4) showing successful Carbon run for HR Feed using `carbon.config.Config` 
    image

6. [Log](https://us-east-1.console.aws.amazon.com/cloudwatch/home?region=us-east-1#logsV2:log-groups/log-group/carbon-ecs-stage/log-events/carbon$252Fcarbon-ecs-stage$252F62ddb26d92ef440fb2945618feae0be5) showing successful Carbon run for Articles Feed using `carbon.config.Config`
   image

7. Screenshot showing XML files on Elements FTP server.
   image

### Includes new or updated dependencies?
NO

### Developer
- [ ] README is updated to reflect all changes as needed
- [ ] All related Jira tickets are linked in commit message(s)

### Code Reviewer
- [ ] The commit message is clear and follows our guidelines
      (not just this pull request message)
- [ ] There are appropriate tests covering any new functionality
- [ ] The documentation has been updated to reflect all changes or is not needed
- [ ] The changes have been verified
- [ ] New dependencies are appropriate or there were no changes

### CODE REVIEW

**Comment by jonavellecuerdo** [comment]
@ghukill @ehanson8 I just added the screenshot showing the successful upload of XML files to the Elements FTP server to 
the PR description.

File: carbon/config.py [hunk 1]
```diff
@@ -1,73 +1,96 @@
...
+    WORKSPACE: str
 
+    def __init__(
+        self,
+        log_level: str = "INFO",
+    ) -> None:
+        self.log_level = log_level
+        self.configure_logger()
+        self.load_environment_variables()
+        self.configure_sentry()
```
Comment (ghukill) [inline carbon/config.py:37]: _Really_ minor here, but as we're talking about the `Config` class...

I might propose a blank line between the attributes passed and set, and these methods that perform initialization logic.  Just a single blank line might be enough to set them apart, e.g.

```python
self.log_level = log_level

self.configure_logger()
self.load_environment_variables()
self.configure_sentry()
```

**Review by ghukill — APPROVED** [review]
I think it looks good!  Nice work.

All my comments are questions / suggestions, so not formally requesting any changes and consider it approved.  If you decide to implement any of these changes happy to re-review, but if you'd prefer to skip, okay with that too!

File: carbon/config.py [hunk 2]
```diff
@@ -1,73 +1,96 @@
...
 
+        if log_level_code < 20:  # noqa: PLR2004
+            logging.basicConfig(
+                format="%(asctime)s %(levelname)s %(name)s.%(funcName)s() "
+                "line %(lineno)d: %(message)s"
+            )
+        else:
+            logging.basicConfig(
+                format="%(asctime)s %(levelname)s %(name)s.%(funcName)s(): %(message)s"
+            )
```
[the remark this answers is not in the document]
Comment (ehanson8) [reply carbon/config.py:56]: Totally agree, a much more understandable approach

File: carbon/config.py (hunk 1 above)
Comment (ehanson8) [reply carbon/config.py:37]: Also agree about this!

File: carbon/app.py [hunk 3]
```diff
@@ -240,3 +263,36 @@ def run_connection_test(self) -> None:
...
+            connect to the Data Warehouse.
+        pipe (DatabaseToFilePipe | DatabaseToFtpPipe): The pipe used to run the
+            data feed. If the pipe is an instance of carbon.app.DatabaseToFtpPipe,
+            a connection test for the Elements FTP server is run.
+    """
+    # test connection to the Data Warehouse
+    try:
+        engine.run_connection_test()
+    except Exception:  # noqa: BLE001
+        logger.error(  # noqa: TRY400
```
Comment (ehanson8) [inline carbon/app.py:284]: Why not use `logger.exception` here to get the traceback as well?

File: carbon/app.py [hunk 4]
```diff
@@ -240,3 +263,36 @@ def run_connection_test(self) -> None:
...
+            "The Data Warehouse connection test failed. The application is exiting."
+        )
+        return
+
+    # test connection to the Symplectic Elements FTP server
+    if isinstance(pipe, DatabaseToFtpPipe):
+        try:
+            pipe.run_connection_test()
+        except Exception:  # noqa: BLE001
+            logger.error(  # noqa: TRY400
```
Comment (ehanson8) [inline carbon/app.py:294]: Same question as above

**Review by ehanson8 — COMMENTED** [review]
Looks good, a few questions. And I would recommend making the changes Graham suggested to streamline it. Great work!

File: carbon/config.py (hunk 1 above)
Comment (jonavellecuerdo) [reply carbon/config.py:37]: Blank line has been added!

File: carbon/app.py (hunk 3 above)
Comment (jonavellecuerdo) [reply carbon/app.py:284]: The reason for this is that the traceback would already be logged at the lower level. 

For `engine.run_connection_test()` the try-except code blocks in lines 113-122 of `carbon.database` make it such that the traceback will be logged for all other errors _except_ `DatabaseError`, which are pretty straight forward. 

This [log](https://us-east-1.console.aws.amazon.com/cloudwatch/home?region=us-east-1#logsV2:log-groups/log-group/carbon-ecs-stage/log-events/carbon$252Fcarbon-ecs-stage$252Ffccf51066550475cbca712465f035304) shows what it would be like if the Data Warehouse connection test failed and `logger.exception` is used at the higher-level (in `carbon.app`). 

The error message is summarized by the fifth line of the log: 
image

If we use `logger.exception`, the log ends with a super long traceback that ultimatley shows the same `DatabaseError` message--rather than a clean message indicating the program is exiting.

---

In short, this code allows us to throw exceptions with traceback at the low-level and cleanly exit the program (i.e. if any of the connection tests fail, the logs will end with: 

>  ERROR carbon.app.run_all_connection_tests(): The Data Warehouse connection test failed. The application is exiting.

Any traceback for non-`DatabaseError` exceptions would appear _before_ this exit message.

---

The same reason applies for your question below! Let me know if you need further clarification.

File: carbon/app.py (hunk 4 above)
Comment (jonavellecuerdo) [reply carbon/app.py:294]: See above!

File: carbon/app.py (hunk 3 above)
Comment (jonavellecuerdo) [reply carbon/app.py:284]: The current code produces a much [cleaner, shorter log](https://us-east-1.console.aws.amazon.com/cloudwatch/home?region=us-east-1#logsV2:log-groups/log-group/carbon-ecs-stage/log-events/carbon$252Fcarbon-ecs-stage$252F37b73d41b7074437b0618a2e2e56d457
) (in the case of `DatabaseError` exceptions): 
image

Comment (ehanson8) [reply carbon/app.py:284]: Got it, I appreciate your clarification!
